### ISSUE TITLE
fori_loop with disable_jit fails when no actual iterations happen

### ISSUE THREAD

**Issue by Jakob-Unfried** [google/jax#8152]
In a trivial call of `fori_loop`, i.e. when `upper == lower`, a normal call works just fine,
but a call with `disable_jit` causes a `TypeError`.

Can be reproduced by 
```python
from jax import lax, disable_jit


def test():
    non_trivial = lax.fori_loop(lower=0, upper=3, body_fun=lambda x, i: x + i, init_val=1)
    print(f'non-trivial result: {non_trivial}')
    trivial = lax.fori_loop(lower=0, upper=0, body_fun=lambda x, i: x + i, init_val=1)
    print(f'trivial result: {trivial}')


test()
# non-trivial result: 4
# trivial result: 1

with disable_jit():
    test()
# non-trivial result: 4
# TypeError: tree_map() missing 1 required positional argument: 'tree'
```

This is because the [non-jit implementation of `lax.scan`](https://github.com/google/jax/blob/cd3d37f4d14bdce8501795c759be7a5b9f79e157/jax/_src/lax/control_flow.py#L1286)
calls `tree_multimap` to stack the outputs, but there aren't any outputs to stack.

I suggest to check for this special case (of a trivial array to scan over) in the non-jit branch of scan and directly return the initial carry and a trivial stacked_y.

Happy to make a PR if this is the way you want to do it.

**Comment by Jakob-Unfried** [comment]
PR see above.
I am not sure if `jax.numpy.array([])` is the suitable stand-in here, or if it should be `None`...

**Comment by jakevdp** [comment]
Hmm... so I think there's one issue with that approach. Look at this for example:
```python
import jax.numpy as jnp
from jax import lax
carry, out = lax.scan(lambda carry, x: (carry, x[None]), 0, jnp.array([]))
print(out.shape)
# (0, 1)
```
I don't think returning empty arrays is precisely correct. You actually do need to know the return type of the function, and construct the output based on that. This implies that in the `length = 0` case with `jit` disabled, we'd have to trace the function to figure out the appropriate output... but disabling `jit` generally means you don't want the function to be traced.

I'm not sure how to resolve this.

**Comment by jakevdp** [comment]
Maybe the safest thing is to raise an explicit error, like
```python
raise ValueError("zero-length scan is not supported in disable_jit() mode because the output type is unknown.")
```

**Comment by jakevdp** [comment]
Closed by #8160

**Comment by mattjj** [comment]
We could handle this case using `jax.eval_shape` or similar.
